libvast/src/system/connector.cpp

```diff
@@ -0,0 +1,168 @@
+  auto err_enum = caf::sec{caf::sec::none};
+  if (!caf::from_integer(err_code, err_enum)) {
+    VAST_WARN("unable to retrieve error code for a remote node connection "
+              "error:{}",
+              err);
+    return false;
+  }
+  return is_non_recoverable_error_enum(err_enum);
+}
+} // namespace
+
+connector_actor::behavior_type
+connector(connector_actor::stateful_pointer<connector_state> self,
+          std::size_t max_connection_attempts,
+          caf::timespan connection_retry_delay, vast::port port,
+          std::string host, caf::timespan connection_timeout) {
+  self->state.connection_retry_delay = connection_retry_delay;
+  self->state.port = std::move(port);
+  self->state.host = std::move(host);
+  self->state.node_connection_timeout = connection_timeout;
+  return {[self,
+           max_connection_attempts](atom::connect) -> caf::result<node_actor> {
+            if (self->state.promise.pending())
+              return self->state.promise;
+            self->state.remaining_connection_attempts = max_connection_attempts;
+            self->state.promise = self->make_response_promise<node_actor>();
+            VAST_INFO("client connects to VAST node at {}", self->state.host);
+            self->send(self, atom::internal_v, atom::start_v);
+            return self->state.promise;
+          },
+          [self](atom::internal, atom::start) {
+            auto middleman = self->system().has_openssl_manager()
+                               ? self->system().openssl_manager().actor_handle()
+                               : self->system().middleman().actor_handle();
+            self
+              ->request(middleman, self->state.node_connection_timeout,
+                        caf::connect_atom_v, self->state.host,
+                        self->state.port.number())
+              .then(
+                [self](const caf::node_id& node_id, caf::strong_actor_ptr& node,
+                       const std::set<std::string>&) {
+                  VAST_DEBUG("client connected to VAST node: {} at {}:{}",
+                             to_string(node_id), self->state.host,
+                             to_string(self->state.port));
+                  self->state.promise.deliver(
+                    caf::actor_cast<node_actor>(std::move(node)));
+                },
+                [self](caf::error& err) {
+                  if (is_non_recoverable_error(err)
+                      || --self->state.remaining_connection_attempts == 0) {
+                    self->state.promise.deliver(caf::make_error(
+                      ec::system_error,
+                      fmt::format("failed to connect to VAST node at {}:{}: {}",
+                                  self->state.host, self->state.port.number(),
+                                  std::move(err))));
+                  } else {
+                    self->delayed_send(self, self->state.connection_retry_delay,
```

Comment:
`self->delayed_send(self, delay, ...)` introduces a hang for the duration of the delay, so this shouldn't be used here. If you wait for #2832 to be merged you can replace this with

```cpp
detail::weak_run_delayed(self, delay, [self] {
  self->send(self, atom::internal_v, atom::start_v);
});
```

Or just replace the internal handler altogether, I don't think it should be needed here at all if you model this with delegations rather a response promise.

Reply:
Is there something i could delegate if i implement it in a single handler?
I want to return node_actor here, but the request to middleman has different API. I can't delegate this message AFAIK.
The success case must convert the output and the error case must trigger retry.

Reply:
That's how I see it as well. The promise is the piece that precisely decouples the two operations. That's what we want. Otherwise retrying is not possible.

Reply:
I'm not arguing against using response promises, I'm arguing against using response promises shared between multiple handlers, i.e., stored in the actor state. These are very hard to reason about correctly and as such make the code less maintainable. We've had a fair amount of issues in the past because of them.

Here's how you can do it without keeping a promise in the state:

```c++
[self](atom::connect) {
  setup()
  // delegate to internal handler
  return self->delegate(self, atom::internal_v, atom::connect_v);
},
[self](atom::internal, atom::connect) {
  auto rp = self->make_response_promise<node_actor>();
  connect().then(
    [self, rp](node_actor& node) mutable {
      rp.deliver(std::move(node));
    },
    [self, rp](caf::error& err) mutable {
      if (!should_retry(err)) {
        rp.deliver(std::move(err));
        return;
      }
      detail::run_weak_delayed(
        self, self->state.retry_delay,
        [self, rp]() mutable {
          rp.delegate(self, atom::internal_v, atom::connect_v);
        });
    },
  return rp;
},
```

This has the big advantage that you no longer need to make sure that the connect handler is only called once, and could actually make the connect handler take the endpoint as an argument now.

The `self` passed to `delegate` might need to be `static_cast<connector_actor>(self)` instead, not sure if that deficiency of CAF was fixed already.

Reply:
If we decide to use two handlers then i think delegate should be used (commented in previous comment as well :P)

Reply:
@dominiklohmann This looks like a lot of complexity. I wouldn't be able to write such code, nor reason about it. Why not try it with a single handler?

```cpp
[self](atom::connect) {
  // Initialize promise exactly once.
  auto return_promise = false;
  if (!self->state.promise) {
    self->state.promise = self->make_response_promise<node_actor>();
    return_promise = true;
  }
  // Try to connect.
  connect_via_middleman(...).then(
    [=](node_actor& node) {
      // We're done.
      self->state.promise.deliver(std::move(node));
    },
    [=](caf::error& err) mutable {
      // Bail out if we can't loop.
      if (!should_retry(err)) {
        rp.deliver(std::move(err));
        return;
      // Back off on failure. If this is an issue with hanging, maybe run_weak_delayed here?
      self->delayed_send(self, delay, atom::connect_v);
    },
  );
  // Return the promise to the user, exactly once.
  if (return_promise)
    return self->state.promise;
  else
    return {};
},
```

This is a simple loop, a single message handler, and easy to reason about.

Reply:
@dominiklohmann I thought about the gymnastics with our `run_weak_delayed` helper a bit more. Shouldn't this something we upstream to CAF? That is, shouldn't `self->delayed_send(self, ...)` work exactly the way you implemented it? I find it a lot more readable when going with CAF messaging defaults, easier to teach, and easier to maintain.

Reply:
> This is a simple loop and easy to reason about.

Your code snippet won't work, as it breaks the promise for concurrent requests. Storing the promise in the state is not a good idea as now you'd need to track the request alongside it. This is a category of errors that you can avoid if you keep the promise to the handler that created it, as I explained [here](https://github.com/tenzir/vast/pull/2835#discussion_r1061512572).

Regarding readability, I think everyone that works with CAF must understand messaging basics, including delegations, so this must not be an argument. We should aim to write maintainable and correct code that is easy to reason about: If a handler can be designed such that it does not need to modify state then we should aim for that. This is a local complexity increase that reduces the complexity of the system.

You can implement the snippet I wrote with a single handler as well, it all comes down to whether you have some setup in the state that you only want to do once. If you don't need that you also don't need a second handler.

I would actually go one further with the approach that I wrote and factor out the business logic from the requests, which also leaves you with just a single handler and even less things that need to be kept in the state:

```cpp
class connection_request {
  // Get the connection request, or an error.
  caf::expected<connection_request> retry(const caf::error& err) const;

  // ...

private:
  std::string endpoint_;
  size_t remaining_retries_;
  caf::timespan retry_delay_;
};

// ...

[self](atom::connect, connection_request req) {
  auto rp = self->make_response_promise<node_actor>();
  connect_via_middleman(...).then(
    [self, rp](const caf::actor& node) mutable {
      rp.deliver(caf::actor_cast<node_actor>(node));
    },
    [self, rp, req = std::move(req)](caf::error& err) mutable {
      auto retry_req = req.retry(err);
      if (!retry_req) {
        rp.deliver(std::move(retry_req.error()));
        return;
      }
      detail::weak_run_delayed(
        self, self->state.retry_delay,
        [self, rp, req = std::move(*retry_req)]() mutable {
          rp.delegate(self, atom::connect_v, std::move(req));
        });
    });
  return rp;
},
```

<hr />

> That is, shouldn't self->delayed_send(self, ...) work exactly the way you implemented it?

Not exactly. The following two do the same thing:

```cpp
// <= CAF 0.17
self->delayed_send(handle, delay, ...)

// >= CAF 0.18
self->run_delayed(delay, [self] { self->send(handle, ...); });
```

But they both keep `self` alive for the duration of the delay because they store a strong reference to `self` on the actor clock.

The `weak_run_delayed` function I wrote stores a weak actor pointer on the actor clock instead and only calls the delayed action if the actor is still alive when the delay expired.

NB: The advantage of the more generic `self->run_delayed(...)` introduced with CAF 0.18 is that it fills the gap of the nonexistent but desirable `self->delayed_delegate(...)`, `rp.delayed_delegate(...)`, and `self->delayed_request(...)`. I consider `self->delayed_send(...)` to be a function that should be deprecated to avoid API bloat.

Reply:
Lot's of details, much appreciated for you taking the time to elaborate.

Overall, I'm in agreement with your solution. It's indeed nicer to basically move the state around, rather than to manually initialize and teardown (same issues as with non-RAII code).

That said, really the only *mutable* state is the counter in this case, everything else can simply be copied. I would say that just wrapping the counter doesn't require shipping a around a "fat" struct that also needs its own type ID. But that's just a minor implementation trade-off. The overall approach is nice.

I would still argue that CAF should provide that "weak" version of `run_delay` that you've implemented.

Reply:
I opened a PR upstream for this, but I don't wanna go through the hoops of backporting it to 0.18.x and ask for a new release for such a minor function. We can live with our own version until then.

> I would say that just wrapping the counter doesn't require shipping a around a "fat" struct that also needs its own type ID. But that's just a minor implementation trade-off.

I think it's worth it for legibility at the call site alone. Sending a `size_t` is much less expressive than sending a `connection_request`. It's a write-once read-many-times effort. Plus I think we actually do want to include the endpoint and the initial retry delay, there's no reason to pass them to the actor on construction instead.

Reply:
I have no strong opinion on the final part of implementation. I'm glad we're agreeing on the *design pattern*. That's what creates long-term sustainability. Whether it's a `size_t` or small struct is just the icing on the cake :wink:.